Updated privacy and security sections

mattgarrish opened the pull request:
This PR takes the work that @wareid did and integrates it into the core and reading system specs.

I don't expect it's perfect at this point, but hopefully we can use the PR to keep refining the sections.

- Reading Systems [preview](https://raw.githack.com/w3c/epub-specs/update/privacy-security/epub33/rs/index.html)
- Reading Systems [diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub33/rs/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://raw.githack.com/w3c/epub-specs/update/privacy-security/epub33/rs/index.html)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/1972.html" title="Last updated on Feb 4, 2022, 12:52 PM UTC (2033c02)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/1972/9738448...2033c02.html" title="Last updated on Feb 4, 2022, 12:52 PM UTC (2033c02)">Diff</a>

mattgarrish committed:
basic port of wendy's work

mattgarrish committed:
Merge remote-tracking branch 'remotes/origin/main' into update/privacy-security

mattgarrish committed:
first pass on core spec

mattgarrish committed:
rough draft of rs section

wareid committed:
filling in missing sections

mattgarrish committed:
add rs draft plus updates

mattgarrish committed:
Merge branch 'update/privacy-security' of https://github.com/w3c/epub-specs into update/privacy-security

mattgarrish committed:
minor fixes

mattgarrish committed:
add bullet for external links

iherman commented on epub33/core/index.html:
```diff
@@ -8773,50 +8773,183 @@ <h2>Accessibility</h2>
+				<h3>Threat Model</h3>
+
+				<p>There are two primary ways that EPUB Publications present risks to users. The first is to trick the
+					user into believing that they are interacting with content from a trustworthy source. These
```
(Editorial) The sentence starts with "The first...", but where is the second? Are the "two primary ways" referring to the two `<dt>/<dd>` pairs below?

mattgarrish replied:
Ya, lost me train of thought there... 😄 

I'm going to rearrange the sections so the deceptions come second. It makes for a more logical introduction to them.

iherman commented on epub33/core/index.html:
```diff
@@ -8773,50 +8773,183 @@ <h2>Accessibility</h2>
+					</dd>
+
+					<dt>Spoofed platforms</dt>
+					<dd>
```
Terminology issue: the term "platform" has not been defined elsewhere in the spec, afaik. I presume we mean the Reading System "platform", the Reading System User interface, or something similar. It would be good to use a more precise terminology. (The term is also used in the description.)

iherman replied:
Reading it again it may not be that important after all. That part is not formal anyway.

iherman commented on epub33/core/index.html:
```diff
@@ -8773,50 +8773,183 @@ <h2>Accessibility</h2>
+					<dt>Embedding of remote resources</dt>
+					<dd>
+						<p>EPUB 3 allows some resources to be <a href="#dfn-remote-resource">remotely hosted</a>,
+							specifically resources whose sizes can negatively affect the downloading and opening of the
```
Probably worth adding an example, like "(e.g., video or font files)". (This section may be read by a different audience who may not immediately understand what we are talking about.)

iherman commented on epub33/core/index.html:
I think that the fact that trigger and switch are now deprecated may need a more emphatic statement than that (just to avoid unnecessary discussions about this). Something like

```
Previous versions of the EPUB specification introduced content switching 
and multimedia control elements, that allow hiding of content and script-less 
control of playback in HTML. These features are now deprecated.
```

We may also want to add, in the Recommendations' section, that authors should _not_ use these features.

(Would it be worth adding, more generally, that authors should not user deprecated features _at all_?)

mattgarrish replied:
The deprecated section says authors should not use the features and they generate warnings in epubcheck, so that keeps most publishers from using them anymore.

I'm not sure about putting them in the recommendations specifically if we're saying there's no threat from them. Maybe what we can do instead is recommend against using _any_ deprecated feature as there could be unknown security or privacy issues due to the lack of any further development of them?

mattgarrish replied:
How about this as an additional bullet for that section:

- Avoiding deprecated features of EPUB due to the potential for undiscovered bugs in implementations.

iherman commented on epub33/core/index.html:
"As described in the thread model": I presume we are talking about the threat model of the RS a this point. If so, let us make it explicit also with a pointer:

> As described in the [threat model of the Reading Systems](https://www.w3.org/TR/epub-rs-33/#epub-threat-model)

mattgarrish replied:
No, that section used to be coequal with the threat model so it was supposed to be a link across. We don't need it at all anymore now that it's a subsection.

iherman commented on epub33/core/index.html:
I know this is controversial, but: should we add some words whereby the content author should only use scripting when strictly necessary for a better consumption of the content and should favor declarative means when possible?

mattgarrish replied:
What if we say: "Not using scripts to send or receive data over the network or store information without the consent of the user."?

That ties into what we're recommending elsewhere, like giving the user control over network activity in the RS spec.

Scaring people off scripting when it only operates on the content seems a bit excessive.

iherman commented on epub33/rs/index.html:
```suggestion
						<p><a>Remote resources</a> present the same risks as any EPUB Publication loaded from an untrusted
```

iherman commented on epub33/rs/index.html:
```suggestion
						data-cite="epub-33#app-media-type"><code>application/epub+zip</code></a>
```

iherman commented on epub33/rs/index.html:
```diff
@@ -1984,49 +2053,83 @@ <h4>Scripting Considerations</h4>
 					maintain the publication’s unique origin when the publication is removed and subsequently
 					re-imported into the content library. Conversely, Reading Systems may create a new unique origin for
-					every newly-added publication.</p>
+					every newly added publication.</p>
```
(The GH review system does not allow me to put a comment to lines on the right text because that was not changed. The comment is on the paragraph talking about the origin.)

There is a reference to §7.1.1 but, I believe, the issues and statements listed in §4.4 should also be referred to,  (see the second and third bullet items after the line "If a Reading System supports scripting" in that section). I am not sure at this point what would be the best way of incorporating this into the current text; maybe it is enough to add an extra reference after §7.1.1 in the same parenthetical statement. Alternatively, something slightly more elaborate would be to add a separate sentence before the one starting with "Examples of Web APIs…":

> This specification also recommends that Reading System assign a unique origin that is shared by all spine-level scripts of the EPUB Publication, and that container-constrained scripts are not allowed to modify the DOM of the parent Content Document and  or manipulate the containing rectangle (see <a href="#sec-scripted-content"></a>).

(Or something similar.)

mattgarrish replied:
Looks like I missed this comment earlier...

What I've done is split the proposed paragraph, and the original paragraph, otherwise the part about container-constrained scripts interrupts the discussion about origin. So the original paragraph now ends with:

"Assigning a unique origin ensures that [spine-level scripts](https://www.w3.org/TR/epub-33/#sec-scripted-spine) [[EPUB-33](http://localhost:120/rs/#bib-epub-33)] are isolated from other EPUB Publications, and limits access to cookies, DOM storage, etc."

The text beginning "Examples of Web APIs that are tied to ..." now starts a new paragraph, and then after that I added the rest of your proposal:

"This specification also recommends that [container-constrained scripts](http://localhost:120/rs/#sec-scripted-container-constrained) [[EPUB-33](http://localhost:120/rs/#bib-epub-33)] not be allowed to modify the DOM of the parent EPUB Content Document and/or manipulate the containing rectangle (see [4.4 Scripting](http://localhost:120/rs/#sec-scripted-content))."

iherman replied:
I do not know what is happening, but I do not find these texts in the previews... :-(

mattgarrish replied:
Ya, we've had this before. It's seems statically.io caches an earlier version and continues to serve it.

Here are alternative links using githack:

- Reading Systems [preview](https://raw.githack.com/w3c/epub-specs/update/privacy-security/epub33/rs/index.html)
- Reading Systems [diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub33/rs/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://raw.githack.com/w3c/epub-specs/update/privacy-security/epub33/rs/index.html)

iherman replied:
Ah... Thanks, found it and it is fine indeed. Thx.

I but you may want to look at the new paragraph "This specification also...": respec does not find the reference for the container-constrainted script and reports a broken reference... (you can consider this issue as resolved once that is handles).

mattgarrish replied:
Ah, ya, that should have been a link across to the other spec. Fixed now.

iherman commented on epub33/core/index.html:
```diff
@@ -8773,50 +8773,183 @@ <h2>Accessibility</h2>
-				CSS, SVG, JavaScript, and other resources — for distribution in a single-file container.</p>
+				<p>This means that EPUB 3's security and privacy issues are primarily linked to the features of those
+					formats, and closely mirror the threats presented by web content generally.</p>
 
```
This is just transferring a note from the WG discussion [on Jan 20](https://www.w3.org/publishing/groups/epub-wg/Meetings/Minutes/2022-01-20-epub#section1): 

> **Wendy Reid:** everything for a given epub is contained within its own package, does not depend on other epubs.
… which is good for privacy, but maybe we could highlight that more.
> 
> **Dan Lazin:** or encourage publishers to make sure that epubs are more self-contained (we allow remote resources).
> … e.g. DAISY accessibility hazards announce to a store that this epub may or may not meet your needs, so we could do similar metadata tagging re. privacy and security risks.

This may be the place to add this remark (and to its counterpart in the other spec.

mattgarrish replied:
This sounds like a useful way to end off the recommendations section. For now, I've added the following paragraph to the end:

"EPUB Creators who want to maximally limit the privacy and security issues in their EPUB Publications should work to make the content as self-contained as possible. An EPUB Publication that comes with all its needed resources and has no dependencies on network access or links to external content not only benefits users but reduces future maintenance and improves archivability."

mattgarrish committed:
add sections for recommending network access and external link practices;
move scripting security back to scripting section;
move local storage recommendations into scripting section;
move event model considerations into scripting section

mattgarrish committed:
Update epub33/rs/index.html

Co-authored-by: Ivan Herman <ivan@w3.org>

mattgarrish committed:
Update epub33/rs/index.html

Co-authored-by: Ivan Herman <ivan@w3.org>

mattgarrish committed:
fixes per Ivan's comments

mattgarrish committed:
Merge branch 'update/privacy-security' of https://github.com/w3c/epub-specs into update/privacy-security

mattgarrish committed:
clarify effect of unique origin for isolating scripts and add para about how the spec recommends isolating container-constrained scripts

mattgarrish committed:
add paragraph about benefits of self-containment

mattgarrish committed:
minor rewording

iherman commented on epub33/core/index.html:
```suggestion
					<p>The one potential exception is the <a data-cite="epub-rs-33#app-epubReadingSystem"
								><code>epubReadingSystem</code> object</a> [[EPUB-RS-33]] that allows EPUB Creators to query
						information about the current Reading System. EPUB Creators need to be mindful that they only
```

mattgarrish committed:
Update epub33/core/index.html

Co-authored-by: Ivan Herman <ivan@w3.org>

mattgarrish committed:
fix link to epub-33 spec

mattgarrish committed:
Merge branch 'update/privacy-security' of https://github.com/w3c/epub-specs into update/privacy-security

mattgarrish committed:
add change log entries for privacy and security changes

mattgarrish committed:
Merge branch 'main' into update/privacy-security
